crates/starknet-types-rpc/src/custom_serde/num_as_hex.rs

```diff
@@ -128,16 +106,20 @@ impl<'de> NumAsHex<'de> for u64 {
                 //
                 // Because we already checked the size of the string earlier, we know that
                 // the following code will never overflow.
+                let hex_bytes = &bytes[2..];
                 let mut n = 0u64;
-                for &b in bytes.iter() {
-                    let unit = match b {
-                        b'0'..=b'9' => b as u64 - b'0' as u64,
-                        b'a'..=b'f' => b as u64 - b'a' as u64 + 10,
-                        b'A'..=b'F' => b as u64 - b'A' as u64 + 10,
+                for &b in hex_bytes {
+                    let digit = match b {
+                        b'0'..=b'9' => b - b'0',
+                        b'a'..=b'f' => 10 + b - b'a',
+                        b'A'..=b'F' => 10 + b - b'A',
                         _ => return Err(E::custom("invalid hexadecimal digit")),
                     };
-
-                    n = n * 16 + unit;
+                    n = n
+                        .checked_mul(16)
+                        .ok_or_else(|| E::custom("integer overflowed 64-bit"))?
+                        .checked_add(digit as u64)
+                        .ok_or_else(|| E::custom("integer overflowed 64-bit"))?;
```

Comment:
Since invalid characters are already checked at this point, and given we're trying to optimize, just a length check before the loop should suffice for overflow checks: 64-bit numbers take at most 16 hexadecimal digits + 2 for the `0x`, so just make sure `hex_bytes.len() <= 16` and use the unchecked variants here.

Reply:
Thank you for your suggestion. I've updated the code to use `skip_while` to skip leading zeros in hexadecimal strings, counting only the significant digits.